ext/standard/math.c

```diff
@@ -95,22 +95,75 @@ static inline double php_intpow10(int power) {
 static inline double php_round_helper(double value, int mode) {
 	double tmp_value;
 
-	if (value >= 0.0) {
-		tmp_value = floor(value + 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (-0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (0.5 + 2 * floor(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (0.5 + 2 * floor(tmp_value/2.0) - 1.0)))
-		{
-			tmp_value = tmp_value - 1.0;
-		}
-	} else {
-		tmp_value = ceil(value - 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (-0.5 + 2 * ceil(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (-0.5 + 2 * ceil(tmp_value/2.0) + 1.0)))
-		{
-			tmp_value = tmp_value + 1.0;
-		}
+	switch (mode) {
+		case PHP_ROUND_HALF_UP:
+			if (value >= 0.0) {
+				tmp_value = floor(value + 0.5);
+			} else {
+				tmp_value = ceil(value - 0.5);
+			}
+			break;
+		case PHP_ROUND_HALF_DOWN:
+			if (value >= 0.0 ) {
+				tmp_value = floor(value + 0.5);
+				if(value + 0.5 == tmp_value) {
+					tmp_value = tmp_value - 1;
+				}
```

Comment:
`tmp_value = ceil(value - 0.5);` is probably more efficient and should return the same result.

Reply:
That is correct, but unfortunately because nature of C's `ceil()` function, values in the range 0 and 0.5 will be rounded to negative zero. This could lead to inconsistent behaviour with PHP's `ceil()` function.

Reply:
> That is correct, but unfortunately because nature of C's ceil() function, values in the range 0 and 0.5 will be rounded to negative zero

ffs - negative zero strikes again. I'm curious: Is there an existing test case for this?

> This could lead to inconsistent behaviour with PHP's ceil() function.

I don't follow here. This is for the `PHP_ROUND_HALF_DOWN` mode which is something entirely different than what `ceil()` odes.

Reply:
If we want to remove negative zero from `round()` function then we should also remove negative zeros from `ceil()` and `floor()` functions to be consistent. And now we allow for negative zero in `ceil()`: https://onlinephp.io/c/28b89

Reply:
I was not arguing in favor of removing negative zero as a possible result. The existence of that was an oversight in my suggestion. Having negative zero allows for the nice symmetry of `round($foo, PHP_ROUND_HALF_DOWN) === -round(-$foo, PHP_ROUND_HALF_DOWN)`.

Reply:
`copysign(ceil(value - 0.5), value)` appears to do the trick. However while doing the research, I came across https://stackoverflow.com/a/4572677/782822 which indicates that our current implementation for `PHP_ROUND_HALF_UP` is not correct either, because it returns `1` for `0.49999999999999994`, instead of `0`: https://3v4l.org/d6AK6

Reply:
Thank you for your suggestion. I changed the implementation using `copysign()`.

Reply:
With the issue I created the implementation will need to change anyway. That issue should likely be fixed before your RFC moves forward.

Reply:
The issue with negative zero has been fixed in your change. Thank you.